internal/unionstore/pipelined_memdb.go

```diff
@@ -0,0 +1,159 @@
+// Copyright 2024 TiKV Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package unionstore
+
+import (
+	"sync"
+	"sync/atomic"
+
+	tikverr "github.com/tikv/client-go/v2/error"
+	"github.com/tikv/client-go/v2/kv"
+	"github.com/tikv/client-go/v2/metrics"
+)
+
+// PipelinedMemDB is a Store which contains
+//   - a mutable buffer for read and write
+//   - an immutable onflushing buffer for read
+//   - like MemDB, PipelinedMemDB also CANNOT be used concurrently
+type PipelinedMemDB struct {
+	// Like MemDB, this RWMutex only used to ensure memdbSnapGetter.Get will not race with
+	// concurrent memdb.Set, memdb.SetWithFlags, memdb.Delete and memdb.UpdateFlags.
+	sync.RWMutex
```

Comment:
`MemDB` already has an internal `RWMutex`, what is this `RWMutex` used to protect? Is it better to group the to be protected fields inside the `RWMutex` like
```
sync.RWMutex  rwLock {
     ...
}
```

Reply:
This mutex is only used to prevent concurrent read/write from upper layer, actually it protects the whole memdb, used as require. It's a patch to the origin memdb, without this embed mutex, this memdb does not satisfy the upper layer interface.

Reply:
Seems to me the only meaning of the rwlock is to provide implement `RLock` and `RUnlock` of the interface. There is no actual usage of `Lock`, right?

Reply:
In most cases there will be no concurrent read/write, may need to remove the internal mutex operations of MemDB in pipelined usages and lock in `PipelinedMemDB.Set`.